File: ruby/pig-latin/pig_latin.rb
module PigLatin


  def self.translate_core(word)
    vowel_characters = ['a', 'e', 'i', 'o', 'u', 'yt', 'xr']
    consonant_characters = ['qu', 'ch', 'thr', 'th', 'squ', 'sch']

    vowel_word = vowel_characters.map { |e| word.scan(/^(#{e})(.*)/) }
                                  .reject { |t| t.empty? }
                                  .flatten
    return vowel_word.join+'ay' unless vowel_word.empty?

    consonant_word = consonant_characters.map { |e| word.scan(/^(#{e})(.*)/) }
                                          .reject { |t| t.empty? }.first
                                          .flatten rescue []
    return consonant_word.reverse.join + 'ay' unless consonant_word.empty?

    word.scan(/^(.)(.*)/).flatten.reverse.join+'ay'
  end

  def self.translate(words)
    words.split(" ").map { |e| translate_core(e) }.join(' ')
  end

end

module BookKeeping
  VERSION = 1
end

File: python/pig-latin/pig_latin.py
import re

def translate(text):
    return ' '.join([translate_core(t) for t in text.split()])
        

def translate_core(text):
    vowel_characters = ['a', 'e', 'i', 'o', 'u', 'yt', 'xr']
    consonant_characters = ['qu', 'ch', 'thr', 'th', 'rh', 'squ', 'sch']

    for vowel in vowel_characters:
        m = re.match(f'^({vowel})(.*)', text)
        if m:
            return m.group() + 'ay'
    
    for consonant in consonant_characters:
        m = re.match(f'^({consonant})(.*)', text)
        m1 = re.match(f'^({consonant})y(.*)', text)
        if m:
            return m.group(2) + m.group(1) + 'ay'
        if m1:
            return 'y' + m1.group(1) + m.group(2) + 'ay'
    
    m = re.match('(.)(.*)', text)
    return m.group(2) + m.group(1) + 'ay'